tokio-util/src/sync/cancellation_token/maybe_dangling.rs

```diff
@@ -0,0 +1,45 @@
+use core::future::Future;
+use core::mem::MaybeUninit;
+use core::pin::Pin;
+use core::ptr;
+use core::task::{Context, Poll};
+
+/// A wrapper type that avoids implicitly asserting the contents are currently valid.
+/// It's safe to contain a dangling reference, as long as it is not used.
+///
+/// # Safety
+///
+/// The inner type must be dropped before the data it references to.
+///
+// Reference
+// See <https://users.rust-lang.org/t/unsafe-code-review-semi-owning-weak-rwlock-t-guard/95706>
+//
+// TODO: replace this with an official solution once RFC #3336 or similar is available.
+// <https://github.com/rust-lang/rfcs/pull/3336>
+#[repr(transparent)]
+pub(crate) struct MaybeDangling<T>(MaybeUninit<T>);
+
+impl<T> Drop for MaybeDangling<T> {
+    fn drop(&mut self) {
+        // Safety: `0` is always initialized.
+        // It may be dangling, but it's the caller responsibility to ensure the data is valid.
```

Comment:
Actually, it should not be dangling at this point.
```suggestion
        // Safety: `0` is always initialized.
```